include/range/v3/utility/reverse_iterator.hpp

```diff
@@ -0,0 +1,203 @@
+               return reverse_iterator(current + n);
+           }
+           RANGES_CXX14_CONSTEXPR reverse_iterator& operator-=(difference_type n)
+           {
+               current += n;
+               return *this;
+           }
+           CONCEPT_REQUIRES(RandomAccessIterator<iterator_type>{})
+           RANGES_CXX14_CONSTEXPR reference operator[](difference_type n) const
+           {
+               return *(*this + n);
+           }
+        };
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator==(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() == y.base();
+        }
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator<(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() > y.base();
+        }
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator!=(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() != y.base();
+        }
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator>(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() < y.base();
+        }
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator>=(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() <= y.base();
+        }
+
+        template <typename It1, typename It2>
+        RANGES_CXX14_CONSTEXPR
+        bool
+        operator<=(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return x.base() >= y.base();
+        }
+
+        template <typename It1, typename It2,
+                  CONCEPT_REQUIRES_(RandomAccessIterator<It1>{} &&
+                                    RandomAccessIterator<It2>{})>
+        RANGES_CXX14_CONSTEXPR
+        iterator_difference_t<It1>
+        operator-(const reverse_iterator<It1>& x, const reverse_iterator<It2>& y)
+        {
+            return y.base() - x.base();
+        }
+
+        template <typename It, CONCEPT_REQUIRES_(RandomAccessIterator<It>{})>
+        RANGES_CXX14_CONSTEXPR
+        reverse_iterator<It>
+        operator+(iterator_difference_t<reverse_iterator<It>> n, const reverse_iterator<It>& x)
+        {
+            return reverse_iterator<It>(x.base() - n);
+        }
+
+        template <typename It, CONCEPT_REQUIRES_(BidirectionalIterator<It>{})>
+        RANGES_CXX14_CONSTEXPR
+        reverse_iterator<It> make_reverse_iterator(It i)
+        {
+            return reverse_iterator<It>(i);
+        }
+
+    }  // namespace v3
+}  // namespace ranges
+#endif  // RANGES_V3_UTILITY_REVERSE_ITERATOR_HPP
```

Comment:
> 1) I really don't like duplicating std functionality. I've done in in isolated places (e.g. some std iterators weren't modeling the Iterator concepts of the Palo Alto report), but I did it grudgingly.

I don't like it either.

> 2) The iterator adaptors live in utility/iterator.hpp
> 3) Most of this verbiage can be replaced with an appropriately-defined cursor together with the basic_iterator utility.

This is libc++'s `reverse_iterator` implementation. I'll write a reverse iterator using `basic_iterator` and a cursor and make sure it passes all libc++ tests. And put that into `utility/iterator.hpp`.

> 4) Don't use iterator_traits. New-style iterators don't define nested pointer or reference type aliases or operator->.
> Ah. If you remove operator-> then you can nuke the weird ranges::addressof. Cool.

Cool, didn't knew about this. I'll remove those then (`addressof` will come back for variant tho).